src/fees/compute_fees.py

```diff
@@ -0,0 +1,606 @@
+        sell orders the limit buy amount is rounded up and for buy orders the limit sell amount is
+        rounded down.
+        """
+        if self.kind == "sell":
+            current_limit_buy_amount = math.ceil(
+                self.limit_buy_amount
+                * Fraction(self.sell_amount, self.limit_sell_amount)
+            )
+            return self.buy_amount - current_limit_buy_amount
+        if self.kind == "buy":
+            current_limit_sell_amount = int(
+                self.limit_sell_amount
+                * Fraction(self.buy_amount, self.limit_buy_amount)
+            )
+            return current_limit_sell_amount - self.sell_amount
+        raise ValueError(f"Order kind {self.kind} is invalid.")
+
+    def raw_surplus(self, fee_policies: list["FeePolicy"]) -> int:
+        """Compute raw surplus of a trade in the surplus token
+        First, the application of protocol fees is reversed. Then, surplus of the resulting trade
+        is computed."""
+        raw_trade = deepcopy(self)
+        for fee_policy in reversed(fee_policies):
+            raw_trade = fee_policy.reverse_protocol_fee(raw_trade)
+        return raw_trade.surplus()
+
+    def protocol_fee(self, fee_policies):
+        """Compute protocol fees of a trade in the surplus token
+        Protocol fees are computed as the difference of raw surplus and surplus."""
+
+        return self.raw_surplus(fee_policies) - self.surplus()
+
+    def surplus_token(self) -> HexBytes:
+        """Returns the surplus token"""
+        if self.kind == "sell":
+            return self.buy_token
+        if self.kind == "buy":
+            return self.sell_token
+        raise ValueError(f"Order kind {self.kind} is invalid.")
+
+    def price_improvement(self, quote: "Quote") -> int:
+        """Compute price improvement
+        For partially fillable orders, rounding is such that the reference for computing price
+        improvement is as if the quote would determine the limit price. That means that for sell
+        orders the quote buy amount is rounded up and for buy orders the quote sell amount is
+        rounded down.
+        """
+        effective_sell_amount = quote.effective_sell_amount(self.kind)
+        effective_buy_amount = quote.effective_buy_amount(self.kind)
+        if self.kind == "sell":
+            current_limit_quote_amount = math.ceil(
+                effective_buy_amount * Fraction(self.sell_amount, effective_sell_amount)
+            )
+            return self.buy_amount - current_limit_quote_amount
+        if self.kind == "buy":
+            current_quote_sell_amount = int(
+                effective_sell_amount * Fraction(self.buy_amount, effective_buy_amount)
+            )
+            return current_quote_sell_amount - self.sell_amount
+        raise ValueError(f"Order kind {self.kind} is invalid.")
+
+    def compute_surplus_fee(self) -> int:
```

Comment:
Although `surplus_fee` is the term the backend uses, i feel it would be clearer if we just name it `total_fee`.

cc @fhenneke

Reply:
Using `total_fee` is fine as well. But I would object that the total fee is a clear concept in our context. But adding more documentation might resolve that.
`surplus_fee` is nice because it is clear that it is a quantity expressed in surplus. What the backend did with that by converting back to sell tokens is a bit unintuitive. The meaning in this code will probably clash with what backend does.
Maybe `surplus_reduction` or something similar might work.

Reply:
let's add at least some comment to clarify what this is computing